File: Prácticas/P3/Ruby/civitas/pkg/civitas-0.0.1/lib/civitas_juego.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "dado.rb"
require_relative "gestor_estados.rb"
require_relative "tablero.rb"
require_relative "jugador.rb"
require_relative "mazo_sorpresas.rb"
require_relative "estados_juego.rb"
require_relative "operaciones_juego.rb"
require_relative "titulo_propiedad.rb"
require_relative "casilla.rb"

module Civitas
  class CivitasJuego
    def initialize(nombres)
      @jugadores = Array.new
      nombres.each do |i|
        @jugadores << Jugador.new(i)
      end
      @gestor_estados = Gestor_estados.new
      @estado = @gestor_estados.estado_inicial
      @indice_jugador_actual = Dado.instance.quien_empieza(@jugadores.length)
      @mazo = MazoSorpresas.new()
      @tablero = Tablero.new(10)
      inicializar_tablero(@mazo)
      inicializar_mazo_sorpresas(@tablero)
    end
    
    private
    def avanza_jugador()
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      posicion_actual = jugador_actual.num_casilla_actual
      tirada = Dado.instance.tirar
      posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)
      casilla = @tablero.get_casilla(posicion_nueva)
      contabilizar_pasos_por_salida(jugador_actual)
      jugador_actual.mover_a_casilla(posicion_nueva)
      casilla.recibe_jugador(@indice_jugador_actual, @jugadores)
      contabilizar_pasos_por_salida(jugador_actual)
    end
    
    def contabilizar_pasos_por_salida(jugador_actual)
      while @tablero.por_salida>0
        jugador_actual.pasa_por_salida
      end
    end
    
    def pasar_turno()
      @indice_jugador_actual = (@indice_jugador_actual+1)%@jugadores.lenght
    end
    
    def inicializar_tablero(mazo)
      @tablero.añade_casilla(Casilla.ini_descanso("Salida"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/Recogidas Nº1", 650, 70, 10, 650, 450)))
      @tablero.añade_casilla(Casilla.ini_sorpresa(mazo, "Sorpresa1"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/Recogidas Nº2", 750, 80, 15, 700, 500)))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/Recogidas Nº3", 550, 60, -10, 600, 350)))
      @tablero.añade_casilla(Casilla.ini_descanso("Parking"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/San Antón Nº1", 500, 50, -10, 150, 250)))
      @tablero.añade_casilla(Casilla.ini_impuesto(350, "Impuesto"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/San Antón Nº2", 600, 70, 10, 300, 350)))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("C/San Antón Nº3", 550, 60, -10, 250, 300)))
      @tablero.añade_casilla(Casilla.ini_descanso("Cárcel"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Paseo del Violón Nº1", 500, 50, -20, 200, 250)))
      @tablero.añade_casilla(Casilla.ini_sorpresa(mazo, "Sorpresa2"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Paseo del Violón Nº2", 800, 70, 10, 500, 400)))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Paseo del Violón Nº3", 600, 60, -15, 350, 350)))
      @tablero.añade_juez()
      @tablero.añade_casilla(Casilla.ini_sorpresa(mazo, "Sorpresa3"))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Cmno de Ronda Nº1", 950, 90, 15, 900, 650)))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Cmno de Ronda Nº2", 850, 80, 10, 800, 600)))
      @tablero.añade_casilla(Casilla.ini_calle(TituloPropiedad.new("Cmno de Ronda Nº3", 1000, 100, 20, 950, 750)))
    end
    
    def inicializar_mazo_sorpresas(tablero)
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PAGARCOBRAR,1000,"Vas por la calle y te encuentras 1000€ ¡ESTAS DE SUERTE!"))
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PAGARCOBRAR,-1000,"¡OH NO! Has perdido 1000€ Este año no podrás pagar Mordor"))
      @mazo.al_mazo(Sorpresa.ini_mov_otra(TipoSorpresa::IRCASILLA,tablero,tablero.num_casilla_carcel,"¡Ya esta bien! Ve a la cárcel directamente"))
      @mazo.al_mazo(Sorpresa.ini_mov_otra(TipoSorpresa::IRCASILLA,tablero,15,"El juez te reclama"))
      @mazo.al_mazo(Sorpresa.ini_mov_otra(TipoSorpresa::IRCASILLA,tablero,5,"Ve corriendo al parking, te esperamos"))
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PORCASAHOTEL,500,"Cobra 500€ por cada casa o cada hotel que poseas"))
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PORCASAHOTEL,-500,"Paga 500€ por cada casa o cada hotel que poseas"))
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PORJUGADOR,500,"Recibes 500€ de cada jugador"))
      @mazo.al_mazo(Sorpresa.ini_resto(TipoSorpresa::PORJUGADOR,-300,"Paga 300€ a cada jugador"))
      @mazo.al_mazo(Sorpresa.ini_exit(TipoSorpresa::SALIRCARCEL,@mazo))
      @mazo.al_mazo(Sorpresa.ini_carcel(TipoSorpresa::IRCARCEL,tablero))
    end
    
    public
    
    def ranking()
      @jugadores = @jugadores.shuffle
    end
    
    def cancelar_hipoteca(ip)
      @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(ip)
    end
    
    def comprar()
      jugador_actual = get_jugador_actual
      num_casilla_actual = jugador_actual.num_casilla_actual
      casilla = @tablero.get_casilla(num_casilla_actual)
      titulo = casilla.titulo_propiedad
      jugador_actual.comprar(titulo)
    end
    
    def construir_casa(ip)
      @jugadores.at(@indice_jugador_actual).construir_casa(ip)
    end
    
    def construir_hotel(ip)
      @jugadores.at(@indice_jugador_actual).construir_hotel(ip)
    end
    
    def final_del_juego()
      salida = false
      @jugadores.each do |i|
        if i.en_bancarrota
          salida = true
        end
      end
      salida
    end
    
    def get_casilla_actual
      @tablero.get_casilla(get_jugador_actual().num_casilla_actual)
    end
    
    def get_jugador_actual
      @jugadores.at(@indice_jugador_actual)
    end
    
    def hipotecar(ip)
      @jugadores.at(@indice_jugador_actual).hipotecar(ip)
    end
    
    def info_jugador_texto()
      @jugadores.at(@indice_jugador_actual).to_s
    end
    
    def salir_carcel_pagando
      @jugadores.at(@indice_jugador_actual).salir_carcel_pagando()
    end
    
    def salir_carcel_tirando
      @jugadores.at(@indice_jugador_actual).salir_carcel_tirando()
    end
    
    def siguiente_paso()
      jugador_actual = get_jugador_actual
      operacion = @gestor_estados.operaciones_permitidas(jugador_actual, @estado)
      case operacion
      when OperacionesJuego::PASARTURNO
        pasar_turno
        siguiente_paso_completado(operacion)
      when OperacionesJuego::AVANZAR
        avanza_jugador
        siguiente_paso_completado(operacion)
      end
      operacion
    end
    
    def siguiente_paso_completado(operacion)
      @estado = @gestor_estados.siguiente_estado(@jugadores.at(@indice_jugador_actual), @estado, operacion)
    end
    
    def vender(ip)
      @jugadores.at(@indice_jugador_actual).vender(ip)
    end
  end
end

File: Prácticas/P3/Java/civitas/src/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author Jose Padial
 */
public class CivitasJuego {
    private int indiceJugadorActual;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private ArrayList<Jugador> jugadores = new ArrayList();
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private Dado dado = Dado.getInstance();
    
    public CivitasJuego(ArrayList<String> nombres){
        for(String n:nombres){
            jugadores.add(new Jugador(n));
        }
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        indiceJugadorActual = dado.quienEmpieza(jugadores.size());
        mazo = new MazoSorpresas();
        inicializarTablero(mazo);
        inicializarMazoSorpresas(tablero);
    }
    
    private void avanzaJugador(){
        Jugador jugadorActual = getJugadorActual();
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = dado.tirar();
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla = tablero.getCasilla(posicionNueva);
        contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posicionNueva);
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        contabilizarPasosPorSalida(jugadorActual);
    }
    
    public boolean cancelarHipoteca(int ip){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
    }
    
    public boolean comprar(){
        Jugador jugadorActual = getJugadorActual();
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        Casilla casilla = tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = casilla.getTituloPropiedad();
        return jugadorActual.comprar(titulo);
    }
    
    public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
    }
    
    private void contabilizarPasosPorSalida(Jugador juagadorActual){
        int aux = tablero.getPorSalida();
        for(int i=0; i<aux; i++)
            juagadorActual.pasaPorSalida();            
    }
    
    public boolean finalDelJuego(){
        for(Jugador j:jugadores){
            if(j.enBancarrota())
                return true;
        }
        return false;
    }
    
    public Casilla getCasillaActual(){
        return tablero.getCasilla(getJugadorActual().getNumCasillaActual());
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public boolean hipotecar(int ip){
        return jugadores.get(indiceJugadorActual).hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        return jugadores.get(indiceJugadorActual).toString();
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, 1000, "Vas por la calle y te encuentras 1000€ ¡ESTÁS DE SUERTE!"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -1000, "¡OH NO! Has perdido 1000€ Este año no podrás pagar Mordor"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, tablero.getCarcel(), "¡Ya esta bien! Ve a la cárcel directamente"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 15, "El juez te raclama"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 5, "Ve corriendo al parking, te esperamos"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, 400, "Cobra 400€ por cada casa o hotel que poseas"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, -400, "Paga 400€ por cada casa o cada hotel que poseas"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, 450, "Recibes 450€ de cada jugador"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, -300, "Paga 300€ a cada jugador"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL, tablero));
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(10);
        tablero.añadeCasilla(new Casilla("Salida"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/Recogidas Nº1", 650, 70, 10, 650, 450)));
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa1"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/Recogidas Nº2", 750, 80, 15, 700, 500)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/Recogidas Nº3", 550, 60, -10, 600, 350)));
        tablero.añadeCasilla(new Casilla("Parking"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/San Antón Nº1", 500, 50, -10, 150, 250)));
        tablero.añadeCasilla(new Casilla(350, "Impuesto"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/San Antón Nº2", 600, 70, 10, 300, 350)));
        tablero.añadeCasilla(new Casilla("Cárcel"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("C/San Antón Nº3", 550, 60, -10, 250, 300)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Paseo del Violón Nº1", 500, 50, -20, 200, 250)));
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa2"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Paseo del Violón Nº2", 800, 70, 10, 500, 400)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Paseo del Violón Nº3", 600, 60, -15, 350, 350)));
        tablero.añadeJuez();
        tablero.añadeCasilla(new Casilla(mazo, "Sorpresa3"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Cmno de Ronda Nº1", 950, 90, 15, 900, 650)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Cmno de Ronda Nº2", 850, 80, 10, 800, 600)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Cmno de Ronda Nº3", 1000, 100, 20, 950, 750)));
    }
    
    private void pasarTurno(){
        indiceJugadorActual = (indiceJugadorActual+1)%jugadores.size();
    }
    
    public ArrayList<Jugador> ranking(){
        ArrayList<Jugador> rank;
        Collections.sort(jugadores);
        rank = jugadores;
        return rank;
    }
    
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    }
    
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
        switch(operacion){
            case PASAR_TURNO:
                pasarTurno();
                siguientePasoCompletado(operacion);
                break;
            case AVANZAR:
                avanzaJugador();
                siguientePasoCompletado(operacion);
                break;
        }
        return operacion;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
    }
    
    public boolean vender(int ip){
        return jugadores.get(indiceJugadorActual).vender(ip);
    }
}
